src/main/java/io/cdap/plugin/couchbase/sink/CouchbaseRecordWriter.java

```diff
@@ -0,0 +1,109 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.cdap.plugin.couchbase.sink;
+
+import com.couchbase.client.java.Bucket;
+import com.couchbase.client.java.Cluster;
+import com.couchbase.client.java.CouchbaseCluster;
+import com.couchbase.client.java.document.JsonDocument;
+import com.google.common.base.Strings;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import io.cdap.plugin.couchbase.OperationType;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.mapreduce.RecordWriter;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import rx.Observable;
+import rx.functions.Func1;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * {@link CouchbaseRecordWriter} writes the job outputs to the Couchbase. Accepts <code>null</code> key,
+ * {@link JsonDocument} pairs but writes only {@link JsonDocument} to the couchbase.
+ */
+public class CouchbaseRecordWriter extends RecordWriter<NullWritable, JsonDocument> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CouchbaseRecordWriter.class);
+  private static final Gson gson = new GsonBuilder().create();
+
+  private final Func1<JsonDocument, Observable<JsonDocument>> operation;
+  private final Cluster cluster;
+  private final Bucket bucket;
+  private final int batchSize;
+  private List<JsonDocument> batch;
+  private int totalCount;
+
+  public CouchbaseRecordWriter(TaskAttemptContext taskAttemptContext) {
+    Configuration conf = taskAttemptContext.getConfiguration();
+    String configJson = conf.get(CouchbaseOutputFormatProvider.PROPERTY_CONFIG_JSON);
+    CouchbaseSinkConfig config = gson.fromJson(configJson, CouchbaseSinkConfig.class);
+
+    this.cluster = CouchbaseCluster.create(config.getNodeList());
+    if (!Strings.isNullOrEmpty(config.getUser()) || !Strings.isNullOrEmpty(config.getPassword())) {
+      cluster.authenticate(config.getUser(), config.getPassword());
+    }
+    this.bucket = cluster.openBucket(config.getBucket());
+    this.batchSize = config.getBatchSize();
+    this.batch = new ArrayList<>();
+    this.totalCount = 0;
+    OperationType operationType = config.getOperationType();
+    this.operation = operationType == OperationType.INSERT ? docToInsert -> bucket.async().insert(docToInsert)
+      : operationType == OperationType.REPLACE ? docToInsert -> bucket.async().replace(docToInsert)
+      : docToInsert -> bucket.async().upsert(docToInsert);
+  }
+
+  @Override
+  public void write(NullWritable key, JsonDocument document) {
+    LOG.trace("RecordWriter write({})", document);
+    batch.add(document);
+    ++totalCount;
+    if (totalCount % batchSize == 0) {
+      flush();
+    }
+  }
+
+  @Override
+  public void close(TaskAttemptContext taskAttemptContext) {
+    flush();
```

Comment:
can this throw exceptions?

Reply:
`flush()` method invocation moved to the 'try' block and  `bucket.close()`, `cluster.disconnect()` moved to the `finally` block.